### DISCUSSION TITLE
Password Reset with AD Backend

### DISCUSSION THREAD

**Discussion by paul-switchnetsys** [authelia/authelia#5819] [Troubleshooting] [answered]
I'll start with the relevant topics / issues I found that matched, or came close to the issue I'm currently experiencing:

-  #1188 
-  #1715 
-  #4199 
-  #5519 

Now to cover off my setup:

I have two AD (2019) servers and Authelia (4.37.5) running in Docker on Debian 11. Authelia is tucked behind Traefik (also running in Docker on the same Debian host and in the same docker network). Actual authentication is working just fine as I have several apps protected by Authelia using either one-factor, two-factor or OIDC (even added Duo push recently). 

The problem I am experiencing is when my users try to reset their password. In the UI I get the red flyout saying that the supplied password does not meet the policy requirements.

Log excerpt:

{"level":"debug","msg":"Check authorization of subject username= groups= ip=172.22.1.12 and object https://auth.example.com/api/reset-password/identity/start (method POST).","time":"2023-08-09T08:47:11+01:00"}
{"level":"debug","method":"POST","msg":"Sending an email to user paulb (email@example.com) to confirm identity for registering a device.","path":"/api/reset-password/identity/start","remote_ip":"172.22.1.12","time":"2023-08-09T08:47:12+01:00"}
{"level":"debug","msg":"Notifier SMTP client attempting connection to mailserver.example.com:25","time":"2023-08-09T08:47:12+01:00"}
{"level":"debug","msg":"Notifier SMTP client connected successfully","time":"2023-08-09T08:47:12+01:00"}
{"level":"debug","msg":"Notifier SMTP server supports STARTTLS (disableVerifyCert: false, ServerName: mailserver.example.com), attempting","time":"2023-08-09T08:47:12+01:00"}
{"level":"debug","msg":"Notifier SMTP STARTTLS completed without error","time":"2023-08-09T08:47:12+01:00"}
{"level":"debug","msg":"Notifier SMTP server supports authentication with the following mechanisms: CRAM-MD5 PLAIN LOGIN DIGEST-MD5","time":"2023-08-09T08:47:12+01:00"}
{"level":"debug","msg":"Notifier SMTP client attempting AUTH PLAIN with server","time":"2023-08-09T08:47:12+01:00"}
{"level":"debug","msg":"Notifier SMTP client authenticated successfully with the server","time":"2023-08-09T08:47:12+01:00"}
{"level":"debug","msg":"Notifier SMTP client attempting to send email body to \"Paul Barnes\" \u003cuser@example.com\u003e","time":"2023-08-09T08:47:12+01:00"}
{"level":"debug","msg":"Notifier SMTP client successfully sent email","time":"2023-08-09T08:47:12+01:00"}
{"level":"debug","msg":"Check authorization of subject username= groups= ip=172.22.1.12 and object https://auth.example.com/reset-password/step2?token=randomtokenstringhere (method GET).","time":"2023-08-09T08:47:16+01:00"}
{"level":"debug","msg":"Check authorization of subject username= groups= ip=172.22.1.12 and object https://auth.example.com/static/js/index.fad9e36b.js (method GET).","time":"2023-08-09T08:47:17+01:00"}
{"level":"debug","msg":"Check authorization of subject username= groups= ip=172.22.1.12 and object https://auth.example.com/static/css/index.40feef90.css (method GET).","time":"2023-08-09T08:47:17+01:00"}
{"level":"debug","msg":"Check authorization of subject username= groups= ip=172.22.1.12 and object https://auth.example.com/locales/en-US/portal.json (method GET).","time":"2023-08-09T08:47:17+01:00"}
{"level":"debug","msg":"Check authorization of subject username= groups= ip=172.22.1.12 and object https://auth.example.com/locales/en/portal.json (method GET).","time":"2023-08-09T08:47:17+01:00"}
{"level":"debug","msg":"Check authorization of subject username= groups= ip=172.22.1.12 and object https://auth.example.com/manifest.json (method GET).","time":"2023-08-09T08:47:17+01:00"}
{"level":"debug","msg":"Check authorization of subject username= groups= ip=172.22.1.12 and object https://auth.example.com/api/reset-password/identity/finish (method POST).","time":"2023-08-09T08:47:17+01:00"}
{"level":"debug","msg":"Check authorization of subject username= groups= ip=172.22.1.12 and object https://auth.example.com/api/configuration/password-policy (method GET).","time":"2023-08-09T08:47:18+01:00"}
{"level":"debug","msg":"Check authorization of subject username= groups= ip=172.22.1.12 and object https://auth.example.com/api/reset-password (method POST).","time":"2023-08-09T08:47:23+01:00"}
{"level":"error","method":"POST","msg":"**unable to update password. Cause: LDAP Result Code 53 \"Unwilling To Perform\**": 0000052D: SvcErr: DSID-031A124C, problem 5003 (WILL_NOT_PERFORM), data 0\n\u0000","path":"/api/reset-password","remote_ip":"172.22.1.12","stack":[{"File":"github.com/authelia/authelia/v4/internal/middlewares/authelia_context.go","Line":65,"Name":"(*AutheliaCtx).Error"},{"File":"github.com/authelia/authelia/v4/internal/handlers/handler_reset_password_step2.go","Line":43,"Name":"ResetPasswordPOST"},{"File":"github.com/authelia/authelia/v4/internal/middlewares/bridge.go","Line":54,"Name":"(*BridgeBuilder).Build.func1.1"},{"File":"github.com/authelia/authelia/v4/internal/middlewares/headers.go","Line":25,"Name":"SecurityHeadersCSPNone.func1"},{"File":"github.com/authelia/authelia/v4/internal/middlewares/headers.go","Line":35,"Name":"SecurityHeadersNoStore.func1"},{"File":"github.com/authelia/authelia/v4/internal/middlewares/headers.go","Line":16,"Name":"SecurityHeaders.func1"},{"File":"github.com/fasthttp/router@v1.4.14/router.go","Line":414,"Name":"(*Router).Handler"},{"File":"github.com/valyala/fasthttp@v1.43.0/http.go","Line":154,"Name":"(*Response).StatusCode"},{"File":"github.com/authelia/authelia/v4/internal/middlewares/strip_path.go","Line":22,"Name":"StripPath.func1.1"},{"File":"github.com/valyala/fasthttp@v1.43.0/server.go","Line":2338,"Name":"(*Server).serveConn"},{"File":"github.com/valyala/fasthttp@v1.43.0/workerpool.go","Line":224,"Name":"(*workerPool).workerFunc"},{"File":"github.com/valyala/fasthttp@v1.43.0/workerpool.go","Line":196,"Name":"(*workerPool).getCh.func1"},{"File":"runtime/asm_amd64.s","Line":1594,"Name":"goexit"}],"time":"2023-08-09T08:47:24+01:00"}

Prior to v4.37.5 the password reset function did work just fine (unfortunately, I cannot remember specifically **which** version I was running at the time I was successfully able to reset passwords).

Here's the kicker, the AD domain is **not** enforcing password complexity:
`PS C:\Users\Administrator.HQ> Get-ADDefaultDomainPasswordPolicy

**ComplexityEnabled           : False**
DistinguishedName           : DC=hq,DC=example,DC=com
LockoutDuration             : 00:30:00
LockoutObservationWindow    : 00:30:00
LockoutThreshold            : 0
MaxPasswordAge              : 90.00:00:00
MinPasswordAge              : 1.00:00:00
MinPasswordLength           : 6
objectClass                 : {domainDNS}
objectGuid                  : 54b9dd50-86d1-4563-9e21-5a04244f8cf1
PasswordHistoryCount        : 4
ReversibleEncryptionEnabled : False
`

Finally, here's the authentication_backend section of configuration.yml:
`authentication_backend:
  password_reset:
    disable: false
  ldap:
    implementation: activedirectory
    url: ldaps://172.22.1.10
    tls:
      skip_verify: true
      minimum_version: TLS1.2
      server_name: midgard.hq.example.com
    base_dn: DC=hq,DC=example,DC=com
    username_attribute: sAMAccountName
    users_filter: (&(|({username_attribute}={input})({mail_attribute}={input}))(objectCategory=person)(objectClass=user)(!userAccountControl:1.2.840.113556.1.4.803:=2)(!pwdLastSet=0))
    groups_filter: (&(member={dn})(|(sAMAccountType=268435456)(sAMAccountType=536870912)))
    group_name_attribute: cn
    mail_attribute: mail
    permit_referrals: true
    display_name_attribute: displayName
    user: CN=Authelia User,OU=Services,DC=hq,DC=example,DC=com`

I'm 99.99% sure that this has worked in the past I just cannot, for the life of me, figure out what - if anything - has changed to stop it working now.

Happy to provide further logs/traces/pcaps if needed.

Ideas anyone?

**Comment by james-d-elliott** [comment]
We have enabled the active directory control which enforces the domain password policy. The error returned from the server is not entirely transparent about the cause. However [3.1.1.5.3.2 Constraints](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/76df255d-dc67-4c1f-adc6-1e5b60021304) points us to [3.1.5.13.7 SamrValidatePassword (Opnum 67)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-samr/c78a7239-f8fc-4a42-bb71-321e897dc046). There **are several** of these policies still enabled, maybe there is a clearer error logged by Active Directory?

ComplexityEnabled : False
DistinguishedName : DC=hq,DC=example,DC=com
LockoutDuration : 00:30:00
LockoutObservationWindow : 00:30:00
LockoutThreshold : 0
**MaxPasswordAge : 90.00:00:00**
**MinPasswordAge : 1.00:00:00**
**MinPasswordLength : 6**
objectClass : {domainDNS}
objectGuid : 54b9dd50-86d1-4563-9e21-5a04244f8cf1
**PasswordHistoryCount : 4**
ReversibleEncryptionEnabled : False

**Reply by paul-switchnetsys** [reply]
Hi James,

Thanks for the response.

I did enable diagnostic logging for the LDAP interface in the registry to see if anything useful cropped up. The only messages of note are:

Internal event: The LDAP server returned an error. 
 
Additional Data 
Error value:
00000003: LdapErr: DSID-0C06061A, comment: Error decrypting ldap message, data 0, v4563 (Event ID 1535)

Followed by:

Internal event: An LDAP client connection was closed because of an error. 
 
Client IP:
172.22.1.202:52648 
 
Additional Data 
Error value:
3 The system cannot find the path specified. 
Internal ID:
c060613 (Event ID 1216)

The client IP is the docker host IP (for reference).

This would suggest that there's something off with the cert. But, if that was the case, wouldn't authentication be affected also?

Also of concern is the 1216 event about not finding the path. Not entirely sure what is going on there.

There isn't much going on in the Security log, I see the authelia user log in successfully with the correct session destroy.

Sadly, there doesn't appear to be much else of use from AD (unless I've missed a point to check - entirely possible!).

Thanks,

**Reply by james-d-elliott** [reply]
How about disabling all of the password policy constraints temporarily? Or using a random generated password?

**Reply by paul-switchnetsys** [reply]
Unfortunately, I tried using a random password (used the generator in BitWarden) which also failed.

I'll try disabling the constraints via GPO and come back to you, though I would have thought that the 'important' policy would have been complexity rather than length/age?

Thanks,

**Reply by james-d-elliott** [reply, marked as answer]
All elements of the policy are separate constraints except for a few which are linked. The complexity one just enforces mixed character classes (alphabetic, numeric, uppercase, symbols, etc). I will double check with my domain tomorrow and get back to you, but relatively certain that version works as intended. Note one of the restrictions is the min age which looks like it's set to 1 day, which means you can't change more than once per day.

**Reply by paul-switchnetsys** [reply]
Thanks James.

Interestingly, now that I've amended the GPO (can't actually disable elements once set, can only modify) so that MinPasswordAge is now **0** and the PasswordHistoryCount is also **0** - i.e. can change immediately once set and no passwords are remembered, I can now change passwords.

Perhaps the original error message "supplied password does not meet the policy requirements" led me down the wrong rabbit hole? (Especially as a 52D indicates a password restriction - including length, complexity or history) Considering that, in most cases, it usually means complexity rules. Though I'd guess it'd be difficult to elaborate on that considering the error message sent back from AD isn't particularly useful in identifying **which** policy isn't being met (typical MS!).

I suspect, in my case, it was history. Now that it's working I can backtrack bits to see which one breaks it and tune from there.

Thanks,

**Reply by james-d-elliott** [reply]
Yeah it's fairly hard to match error messages from BER LDAP responses to specific reasons. However 52D is a generic one that encompasses any complexity requirement as part of the LDAP standards which is what we look for in showing that message. Microsoft just has named a single one of those "Complexity" which is only on element the 52D can be returned for.

We could directly return the text of the error, but as you see here in the logs which does this, it doesn't actually help, also depending on the area in which we show the actual error message to the user/attacker it could divulge information useful to an attack.

**Reply by paul-switchnetsys** [reply]
This is why I'm not a developer ;-)
